compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -1806,12 +1806,11 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
             output,
             c_variadic,
             implicit_self: decl.inputs.get(0).map_or(hir::ImplicitSelfKind::None, |arg| {
-                let is_mutable_pat = match arg.pat.kind {
-                    PatKind::Ident(BindingMode::ByValue(mt) | BindingMode::ByRef(mt), _, _) => {
-                        mt == Mutability::Mut
-                    }
-                    _ => false,
-                };
+                let is_mutable_pat = matches!(
+                    arg.pat.kind,
+                    PatKind::Ident(BindingMode::ByValue(Mutability::Mut), ..)
+                        | PatKind::Ident(BindingMode::ByRef(Mutability::Mut), ..)
```

Comment:
This can also be 

```suggestion
                    PatKind::Ident(BindingMode::ByValue(Mutability::Mut) | BindingMode::ByRef(Mutability::Mut), ..)
```

rigth?

Reply:
Yes, that's what I started out with. The line gets too long and the formatter turns it into:
```suggestion
                    PatKind::Ident(
                        BindingMode::ByValue(Mutability::Mut) | BindingMode::ByRef(Mutability::Mut),
                        ..
                    )
```
Wasn't entirely happy with that and ended up with what I proposed. If you'd prefer the four line version, that's also fine with me.

Reply:
Hmm.. right... Maybe we should just keep the original match then, it is very straightforward.

Reply:
There's also the option of:
```suggestion
                    PatKind::Ident(ByValue(Mutability::Mut) | ByRef(Mutability::Mut), ..)
```
which would need `use BindingMode::{ByRef, ByValue};` directly above the hunk.

Reply:
Sold :hammer: 

I like it, let's do that

Reply:
Pushed.